src/main/java/frc/robot/swerve/SwerveDrive.java

```diff
@@ -0,0 +1,275 @@
+            INSTANCE = new SwerveDrive(isReal, driveIds, angleIds, encoderIds);
+        }
+    }
+
+    /**
+     * Updates the offset for the gyro.
+     *
+     * @param angle The desired angle. [rad]
+     */
+    public void resetGyro(double angle) {
+        gyro.resetGyro(angle);
+    }
+
+    public void resetGyro() {
+        resetGyro(0);
+    }
+
+    /**
+     * Gets the raw yaw reading from the gyro.
+     *
+     * @return Yaw angle reading from gyro. [rad]
+     */
+    public double getRawYaw() {
+        return loggerInputs.rawYaw;
+    }
+
+    /**
+     * Gets the yaw reading from the gyro with the calculated offset.
+     *
+     * @return Yaw angle with offset. [rad]
+     */
+    public double getYaw() {
+        return loggerInputs.yaw;
+    }
+
+    public double getPitch() {
+        return loggerInputs.pitch;
+    }
+
+    /**
+     * Sets the module states to the desired module states.
+     *
+     * @param desiredModuleStates The desired module states to set the modules to.
+     */
+    public void setModuleStates(SwerveModuleState[] desiredModuleStates) {
+        loggerInputs.desiredModuleStates = Utils.swerveModuleStatesToArray(desiredModuleStates);
+    }
+
+    public void updateModulePositions() {
+        for (int i = 0; i < modulePositions.length; i++) {
+            modulePositions[i] = modules[i].getModulePosition();
+        }
+    }
+
+    public Pose2d getBotPose() {
+        return Utils.arrayToPose2d(loggerInputs.botPose);
+    }
+
+    public SwerveDriveKinematics getKinematics() {
+        return kinematics;
+    }
+
+    public double getVelocity() {
+        return loggerInputs.linearVelocity;
+    }
+
+    public ChassisSpeeds getCurrentSpeeds() {
+        return Utils.arrayToChassisSpeeds(loggerInputs.currentSpeeds);
+    }
+
+    public void resetPose(Pose2d pose) {
+        odometry.resetPosition(new Rotation2d(getYaw()), modulePositions, pose);
+    }
+
+    public void resetPose() {
+        odometry.resetPosition(new Rotation2d(getYaw()), modulePositions, new Pose2d());
+    }
+
+    public boolean encodersConnected() {
+        boolean connected = true;
+        for (int i = 0; i < 4; i++) {
+            connected &= modules[i].encoderConnected();
+        }
+        return connected;
+    }
+
+    public void checkSwerve() {
+        for (int i = 0; i < 4; i++) {
+            modules[i].checkModule();
+        }
+    }
+
+    public void stop() {
+        for (int i = 0; i < 4; i++) {
+            modules[i].setModuleState(new SwerveModuleState(0, modules[i].getModuleState().angle));
+        }
+    }
+
+    public void lock() {
+        modules[0].setModuleState(new SwerveModuleState(0, Rotation2d.fromDegrees(45)));
+        modules[1].setModuleState(new SwerveModuleState(0, Rotation2d.fromDegrees(135)));
+        modules[2].setModuleState(new SwerveModuleState(0, Rotation2d.fromDegrees(315)));
+        modules[3].setModuleState(new SwerveModuleState(0, Rotation2d.fromDegrees(225)));
```

Comment:
Shouldn't `225` come before `315`? If yes, then this can be shortened:
```suggestion
    for (int i = 0; i < modules.length; i++) {
        modules[i].setModuleState(new SwerveModuleState(0, Rotation2d.fromDegrees(i * 45)));
    }
```

Reply:
No, the fourth module needs to be at 225 degrees

Reply:
OK, you're right.

Also attaching my beautiful MS Paint diagram which I used to understand this.
![image](https://github.com/Galaxia5987/common/assets/31829093/16589784-a8e8-4157-bcf8-9e9f5f2d45d3)

Reply:
Well actually the x axis faces upwards, and the y faces left. That way all the wheels go into the swerve, and it doesn't rotate or move linearly.

Reply:
Good to know, but, not sure how that's relevant to the locking. When locking all wheels point inside regardless of the axes directions.